### PR TITLE
linked-list: Issues #389. Add special Python functions __iter__ and _…

### PR DESCRIPTION
**Pull request by leec1979**
…_len__ along with their tests. Tests are skipped by default so that the new functionality is treated as optional extra work, as per the issue.

### CODE REVIEW

File: exercises/linked-list/example.py [hunk 1]
```diff
@@ -45,3 +45,21 @@ def unshift(self, value):
             new_node.next = self.head
             self.head.prev = new_node
             self.head = new_node
+
+    def __len__(self):
```
Comment (rootulp) [inline exercises/linked-list/example.py]: I think this implementation of `__len__` is cleaner than mine #408 but it requires O(n) to run when a call to `__len__` on a linked list could be O(1). Fortunately, this exercise doesn't mention anything about minimizing algorithmic complexity.

File: exercises/linked-list/example.py [hunk 2]
```diff
@@ -45,3 +45,21 @@ def unshift(self, value):
...
+
+        return count
+
+    def __iter__(self):
+        current = self.head
+        while current is not None:
+            yield current
+            current = current.next
+
+        return
```
Comment (rootulp) [inline exercises/linked-list/example.py]: Why the final `return`?

File: exercises/linked-list/linked_list_test.py [hunk 3]
```diff
@@ -44,6 +44,26 @@ def test_all(self):
...
+        self.assertEqual(2, len(self.list))
+
+    @unittest.skip("advanced")
+    def test_iter(self):
+        values = [10, 20, 30, 40]
+        for value in values:
+            self.list.push(value)
+
+        for index, node in enumerate(self.list):
+            self.assertEqual(values[index], node.value)
```
Comment (rootulp) [inline exercises/linked-list/linked_list_test.py]: Why are we making this assertion? It seems unrelated to testing the functionality of `__iter__`

Comment (leec1979) [inline exercises/linked-list/linked_list_test.py]: It's testing that iterating over the linkedlist matches the values in the original list used to populate the linkedlist. Possibly a cleaner way to write that test would have been:

self.assertEqual(values, [x.value for x in self.list])

It's redundant due to the explicit iterator tests that follow but I forgot to remove it.

File: exercises/linked-list/example.py (hunk 2 above)
Comment (leec1979) [inline exercises/linked-list/example.py]: That's a hangover from writing in other languages recently.

File: exercises/linked-list/example.py (hunk 1 above)
Comment (leec1979) [inline exercises/linked-list/example.py]: Thanks for the review. I didn't spot your pull request until after I'd submitted mine - I'm still getting used to github.

My approach is more 'self contained' but worse from a big O perspective. From a teaching perspective it would be good to include both and document the pros/cons of each approach.

Comment (rootulp) [inline exercises/linked-list/example.py]: I definitely agree on your approach being more self contained

File: exercises/linked-list/example.py (hunk 2 above)
Comment (rootulp) [inline exercises/linked-list/example.py]: Sounds like you can remove it :)

File: exercises/linked-list/linked_list_test.py (hunk 3 above)
Comment (rootulp) [inline exercises/linked-list/linked_list_test.py]: Also safe to remove this

**Comment by behrtam** [comment]
Thanks for your work!

Unfortunately #408 already implemented those two methods. As it was earlier and conflict free I merged that PR instead.
